cli/command/stack/kubernetes/ps.go

```diff
@@ -29,78 +37,72 @@ func RunPS(dockerCli *KubeCli, options options.PS) error {
-		return fmt.Errorf("nothing found in stack: %s", namespace)
+	stackName := options.Namespace
+	_, err = stacks.Get(stackName)
+	if apierrs.IsNotFound(err) {
+		return fmt.Errorf("nothing found in stack: %s", stackName)
 	}
-
-	pods, err := fetchPods(namespace, podsClient)
 	if err != nil {
 		return err
 	}
-
+	pods, err := fetchPods(stackName, client.Pods(), filters)
+	if err != nil {
+		return err
+	}
 	if len(pods) == 0 {
-		return fmt.Errorf("nothing found in stack: %s", namespace)
+		return fmt.Errorf("nothing found in stack: %s", stackName)
 	}
+	return printTasks(dockerCli, options, stackName, client, pods)
+}
 
+func printTasks(dockerCli command.Cli, options options.PS, namespace string, client corev1.NodesGetter, pods []apiv1.Pod) error {
 	format := options.Format
-	if len(format) == 0 {
+	if format == "" {
 		format = task.DefaultFormat(dockerCli.ConfigFile(), options.Quiet)
 	}
-	nodeResolver := makeNodeResolver(options.NoResolve, client.Nodes())
 
 	tasks := make([]swarm.Task, len(pods))
 	for i, pod := range pods {
 		tasks[i] = podToTask(pod)
 	}
-	return print(dockerCli, namespace, tasks, pods, nodeResolver, !options.NoTrunc, options.Quiet, format)
-}
-
-type idResolver func(name string) (string, error)
-
-func print(dockerCli command.Cli, namespace string, tasks []swarm.Task, pods []apiv1.Pod, nodeResolver idResolver, trunc, quiet bool, format string) error {
 	sort.Stable(tasksBySlot(tasks))
 
 	names := map[string]string{}
 	nodes := map[string]string{}
 
-	tasksCtx := formatter.Context{
-		Output: dockerCli.Out(),
-		Format: formatter.NewTaskFormat(format, quiet),
-		Trunc:  trunc,
-	}
-
+	cnodes := client.Nodes()
 	for i, task := range tasks {
-		nodeValue, err := nodeResolver(pods[i].Spec.NodeName)
+		nodeValue, err := resolveNode(pods[i].Spec.NodeName, cnodes, options.NoResolve)
 		if err != nil {
 			return err
 		}
-
 		names[task.ID] = fmt.Sprintf("%s_%s", namespace, pods[i].Name)
 		nodes[task.ID] = nodeValue
 	}
 
+	tasksCtx := formatter.Context{
+		Output: dockerCli.Out(),
+		Format: formatter.NewTaskFormat(format, options.Quiet),
+		Trunc:  !options.NoTrunc,
+	}
+
 	return formatter.TaskWrite(tasksCtx, tasks, names, nodes)
 }
 
-func makeNodeResolver(noResolve bool, nodes corev1.NodeInterface) func(string) (string, error) {
+func resolveNode(name string, nodes corev1.NodeInterface, noResolve bool) (string, error) {
 	// Here we have a name and we need to resolve its identifier. To mimic swarm behavior
 	// we need to resolve the id when noresolve is set, otherwise we return the name.
 	if noResolve {
-		return func(name string) (string, error) {
-			n, err := nodes.List(metav1.ListOptions{
-				FieldSelector: fields.OneTermEqualSelector(api.ObjectNameField, name).String(),
-			})
-			if err != nil {
-				return "", err
-			}
-			if len(n.Items) != 1 {
-				return "", fmt.Errorf("could not find node '%s'", name)
-			}
-			return string(n.Items[0].UID), nil
+		n, err := nodes.List(metav1.ListOptions{
```

Comment:
We're going to call this function for every single task, which is going to be super slow with large amounts of tasks. Is there any way we could just do a single `nodes.List` call?

Reply:
Ah yes, well spotted… I've changed the code to request the nodes only once and loop on the tasks afterwards.